### PR TITLE
✨Add allow-blocked-end-date flag and e2e test

### PR DESCRIPTION
**Pull request by cvializ**
Implements #17950 

/to @cathyxz @sparhami 
/cc @estherkim

### CODE REVIEW

File: build-system/tasks/e2e/functional-test-controller.js [hunk 1]
```diff
@@ -44,6 +44,21 @@ class ElementHandle {
   }
 }
 
+/**
+ * Keys to send to the FunctionalTestController#type method to trigger
+ * actions instead of text.
+ * @enum {string}
+ */
+export const Keys = {
```
Comment (estherkim) [inline build-system/tasks/e2e/functional-test-controller.js]: We import `Key` from selenium-webdriver in `selenium-webdriver-controller.js`. Can that be used instead?

Comment (cvializ) [reply build-system/tasks/e2e/functional-test-controller.js]: This is part of the abstraction layer so we can use the same keys when using Selenium as our webdriver and when using Puppeteer without rewriting the tests. The `Key` from `selenium-webdriver` wouldn't work for Puppeteer.

This mapping is also good because it lets us use the browser-standard names from https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key and selenium uses non-standard values

https://github.com/SeleniumHQ/selenium/blob/df40c28b41d4b3953f90eaff84838a9ac052b84a/javascript/node/selenium-webdriver/lib/input.js#L44-L57

Comment (estherkim) [reply build-system/tasks/e2e/functional-test-controller.js]: Okay I see!

File: build-system/app-test.js [hunk 2]
```diff
@@ -34,6 +34,28 @@ function appTestEndpoints(app) {
       ],
     });
   });
+
+  app.use('/date-picker/config.json', (req, res) => {
```
Comment (sparhami) [inline build-system/app-test.js]: I think @cathyxz split some non test app stuff out into separate files, which I think is a pattern we should follow so that this file doesn't end up being huge as well.

File: build-system/tasks/e2e/functional-test-controller.js (hunk 1 above)
Comment (sparhami) [inline build-system/tasks/e2e/functional-test-controller.js]: This enum name should be singular (i.e. `Key`)

File: extensions/amp-date-picker/0.1/amp-date-picker.js [hunk 3]
```diff
@@ -1247,22 +1257,28 @@ export class AmpDatePicker extends AMP.BaseElement {
...
       this.iterateDateRange_(startDate, endDate, index => {
         if (this.blocked_.contains(index)) {
-          containsBlocked = true;
+          blockedCount++;
         }
       });
     }
 
-    if (containsBlocked) {
+    if (blockedCount == 1 && this.allowBlockedEndDate_) {
```
Comment (sparhami) [inline extensions/amp-date-picker/0.1/amp-date-picker.js]: function is getting a bit long, would consider extracting out this logic and the above into a function that just checks for blocked days, since it seems to be orthogonal to the rest of the function.

File: extensions/amp-date-picker/0.1/date-picker-common.js [hunk 4]
```diff
@@ -94,13 +94,34 @@ export function withDatePickerCommon(WrappedComponent) {
...
+   * @param {?moment} endDate
+   * @param {boolean} allowBlockedEndDate
+   * @param {!moment} day
+   */
+  function isDayBlocked(list, startDate, endDate, allowBlockedEndDate, day) {
+    const isBlocked = datesListContains(list, day);
+
+    if (startDate && !endDate && allowBlockedEndDate) {
+      return isBlocked && !isSameDay(day, list.firstDateAfter(startDate));
+    } else {
```
Comment (sparhami) [inline extensions/amp-date-picker/0.1/date-picker-common.js]: drop the else, and unindent

File: extensions/amp-date-picker/0.1/test-e2e/test-blocked-dates.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      const todayCell = await selectDate(controller, ID, today);
+      const nextWeekCell = await selectDate(controller, ID, nextWeek);
+
+      await controller.click(todayCell);
+      await expect(controller.getElementAttribute(todayCell, 'class'))
+          .to.contain('CalendarDay__selected_start');
+
+      // This sequence of commands selects the date one week after the start
+      // date since the initial focused cell is the date after the start date.
+      await controller.type(null, Keys.ArrowLeft);
```
Comment (sparhami) [inline extensions/amp-date-picker/0.1/test-e2e/test-blocked-dates.js]: in tests, not a good idea to rely on the current date. Tests should be (to the extent possible) completely reproducible. For example, your test at some point may go into the next month with the next arrow or down arrow key. It could end up that your test fails at some day in the future where this does not work. While it may work fine in this case, you should use specific dates to test that it works correctly (if that is indeed inspected).

Comment (cvializ) [reply extensions/amp-date-picker/0.1/test-e2e/test-blocked-dates.js]: I agree with you generally, but for a few reasons I think for now this is the way to go:

1. E2E tests are designed to execute with as minimal mocking as possible, and spoofing the date in the test is a kind of mocking
2. There is currently no API in Puppeteer or Selenium to override the date and would have to be implemented
3. The date picker changes the date it opens to depending on the current day and this tests that.
4. This sequence of keys is designed to work for any day.

For these reasons I would prefer to keep current behavior at this time.

File: build-system/app-test.js (hunk 2 above)
Comment (cvializ) [reply build-system/app-test.js]: I'm not sure which non test app stuff you're referring to, since that's what I created this file for, but I'd be glad to do something cleaner in a followup PR.

**Comment by cvializ** [comment]
PTAL

File: build-system/app-test.js (hunk 2 above)
Comment (sparhami) [reply build-system/app-test.js]: Like this:

https://github.com/ampproject/amphtml/blob/98000ef8c85e92def6354a5b40c15ce41315f60b/build-system/app.js#L51

Basically, move the route logic into a separate file, then just import it from `app-test`, to keep this file from growing.

Comment (cvializ) [reply build-system/app-test.js]: Nice that looks pretty good. I will definitely do that in a followup PR

**Comment by cvializ** [comment]
Thanks for the review 🎉

**Comment by selfcatering** [comment]
Just so you know the flag breaks page validation.

**Comment by cvializ** [comment]
Thanks for reporting, we'll have that fixed shortly
